Instructions + scripts for running tests repeatedly

julenka opened the pull request:
## Overview
We recently fixed several intermittent test failures #6058, which were difficult to fix because there isn't an easy way to run tests repeatedly in Unity PlayMode. This PR adds a script + instructions (in UnitTests.md) for how to modify tests to run test multiple times.

I used this script to run each intermittently failing test 100 times and verify my fixes.

## Changes
- Fixes: #6114

## Verification
Please check out the PR, see if you can follow instructions to run a test of your choice 100 times in a row.

> This optional section is a place where you can detail the specific type of verification 
> you want from reviewers. For example, if you want reviewers to checkout the PR locally
> and validate the functionality of specific scenarios, provide instructions
> on the specific scenarios and what you want verified.
>
> If there are specific areas of concern or question feel free to highlight them here so
> that reviewers can watch out for those issues.
>
> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

julenka commented:
Reviewers, the UnitTests.md file has several changes. You can check out a version without markup changes here: https://github.com/microsoft/MixedRealityToolkit-Unity/blob/f1fcd184d0ac9ef46505e3413fef366c075eff9b/Documentation/Contributing/UnitTests.md

julenka commented on scripts/test/generate_repeat_tests.py:
```diff
@@ -0,0 +1,21 @@
+    return "[UnityTest]\npublic IEnumerator A%i%s%i(){ yield return %s();}" % ( i, s, j, s)
+
+def print_tests(testnames, repeat):
+    for i in range(1,repeat):
```
```suggestion
    for i in range(1, repeat):
```

julenka commented on scripts/test/generate_repeat_tests.py:
```diff
@@ -0,0 +1,21 @@
+
+def main(argv):
+    import argparse
+    parser = argparse.ArgumentParser(description="Print code to paste in that repeats tests")
```
```suggestion
    parser = argparse.ArgumentParser(description="Prints code that you can copy / paste into your test file to run tests repeatedly")
```

thalbern commented on Documentation/Contributing/UnitTests.md:
```diff
@@ -29,22 +27,60 @@ Run the tests on the project located at H:\mrtk.dev, with Unity 2018.4.1f1, outp
+
+First, make sure you have [Python 3.X installed](https://www.python.org/downloads/).
+
+Let's say you have a test that you want to run many times:
```
```suggestion
For a single test that needs to me executed multiple times:
```

thalbern commented on Documentation/Contributing/UnitTests.md:
```diff
@@ -29,22 +27,60 @@ Run the tests on the project located at H:\mrtk.dev, with Unity 2018.4.1f1, outp
 
-## Writing Tests for your code
-To ensure MRTK being a stable and reliable toolkit, every feature should come with unit tests and sample usage in one of the example scenes.
+Run the following from a command line (we recommend [PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6#powershell-core))
```
```suggestion
Run the following from a command line ([PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6#powershell-core) is recommended)
```

thalbern commented on Documentation/Contributing/UnitTests.md:
````diff
@@ -29,22 +27,60 @@ Run the tests on the project located at H:\mrtk.dev, with Unity 2018.4.1f1, outp
+```
 
-Having good test coverage in a big codebase like MRTK is crucial for stability and having confidence when doing changes in code.
+Copy and paste the output into your test file. If you need to run multiple tests in sequence:
````
```suggestion
Copy and paste the output into your test file. The following script is for running multiple tests in sequence:
```

thalbern commented on Documentation/Contributing/UnitTests.md:
````diff
@@ -29,22 +27,60 @@ Run the tests on the project located at H:\mrtk.dev, with Unity 2018.4.1f1, outp
+python .\generate_repeat_tests.py -n 5 -t MyTest MySecondTest
+```
+
+Your new test file should now contain
````
```suggestion
The new test file should now contain
```

thalbern commented on Documentation/Contributing/UnitTests.md:
````diff
@@ -29,22 +27,60 @@ Run the tests on the project located at H:\mrtk.dev, with Unity 2018.4.1f1, outp
+public IEnumerator MyTest() {...}
+```
+
+Open the test runner and you should see your new tests that you can call repeatedly.
````
```suggestion
Open the test runner and observe the new tests that can now be called repeatedly.
```

thalbern commented on Documentation/Contributing/UnitTests.md:
```diff
@@ -29,22 +27,60 @@ Run the tests on the project located at H:\mrtk.dev, with Unity 2018.4.1f1, outp
+
+Open the test runner and you should see your new tests that you can call repeatedly.
+
+## Writing Tests
```
```suggestion
## Writing tests
```

thalbern commented on Documentation/Contributing/UnitTests.md:
```diff
@@ -29,22 +27,60 @@ Run the tests on the project located at H:\mrtk.dev, with Unity 2018.4.1f1, outp
+Open the test runner and you should see your new tests that you can call repeatedly.
+
+## Writing Tests
+To ensure MRTK being a stable and reliable toolkit, every feature should come with unit tests and sample usage in one of the example scenes. Having good test coverage in a big codebase like MRTK is crucial for stability and having confidence when doing changes in code.
```
this is explaining our test setup - i would move that to the top of the page bc it's prerequisite for going through your instructions (by then you should already know that we're using unity test runner)

CDiaz-MS commented:
@julenka  I think this doc update is based on an older version of UnitTests.md.  I updated UnitTests.md in #5947 specifically the PlayMode Tests template.

julenka commented on Documentation/Contributing/UnitTests.md:
```diff
@@ -1,21 +1,19 @@
-script located at `Scripts\test\run_playmode_tests.ps1`. This will run the playmode tests exactly as they are executed
-on github / CI (see below), and print results. Here are some examples of how to run the script
+### Running tests from command line
+You can also run the [powershell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6) script located at `Scripts\test\run_playmode_tests.ps1`. This will run the playmode tests exactly as they are executed on github / CI (see below), and print results. Here are some examples of how to run the script
```
Oh, I didn't know about that! I never actually tested that experience. I suppose you could just run that command line tests multiple times and parse output. I think the approach of modifying the code is still faster since every time you run a test from cmd unity does a bunch of extra work. I can mention the command line arguments though.

Troy-Ferrell replied:
Exactly they have different benefits. The python script is great to target one test case and get multiple results in editor but you have to run and copy and paste and then delete/cleanup

Where a command line of run everything 5-10 times while I go to lunch involves less setup and covers more bases for that need

julenka replied:
Done, see run_repeat_tests.ps1

julenka commented:
> @julenka I think this doc update is based on an older version of UnitTests.md. I updated UnitTests.md in #5947 specifically the PlayMode Tests template.

Yeah I hadn't merged your latest changes in. Fixed!

julenka commented:
/azp run
